test/plug_test.exs

```diff
@@ -1542,4 +1548,131 @@ defmodule PlugTest do
+      },
+      :get,
+      "/posts/create",
+      params
+    )
+    expected = Plug.Conn.assign(conn, :authorized, true)
+
+    assert authorize_controller(conn, opts) == expected
+
+    # when the action is "index"
+    params = %{}
+    conn = conn(
+      %Plug.Conn{
+        private: %{phoenix_action: :index, phoenix_controller: Myproject.SampleController},
+        assigns: %{current_user: %User{id: 1}}
+      },
+      :get,
+      "/posts",
+      params
+    )
+    expected = Plug.Conn.assign(conn, :authorized, true)
+
+    assert authorize_controller(conn, opts) == expected
+
+    # when the action is a phoenix action
+    params = %{"id" => 1}
+    conn = conn(
+      %Plug.Conn{
+        private: %{phoenix_action: :show, phoenix_controller: Myproject.SampleController},
+        assigns: %{current_user: %User{id: 1}}
+      },
+      :get,
+      "/posts/1",
+      params
+    )
+    expected = Plug.Conn.assign(conn, :authorized, true)
+
+    assert authorize_controller(conn, opts) == expected
+
+    # when the current user can access the given resource
+    # and the action and controller are specified in conn.assigns.canary_action
+    params = %{"id" => 1}
+    conn = conn(
+      %Plug.Conn{
+        private: %{},
+        assigns: %{current_user: %User{id: 1}, canary_action: :show, canary_controller: Myproject.SampleController}
+      },
+      :get,
+      "/posts/1",
+      params
+    )
+    expected = Plug.Conn.assign(conn, :authorized, true)
+
+    assert authorize_controller(conn, opts) == expected
+
+    # when both conn.assigns.canary_action and conn.private.phoenix_action are defined
+    # it uses conn.assigns.canary_action for authorization
+    params = %{"id" => 1}
+    conn = conn(
+      %Plug.Conn{
+        private: %{phoenix_action: :show, phoenix_controller: Myproject.SampleController},
+        assigns: %{current_user: %User{id: 1}, canary_action: :unauthorized}
+      },
+      :get,
+      "/posts/1",
+      params
+    )
+    expected = Plug.Conn.assign(conn, :authorized, false)
+
+    assert authorize_controller(conn, opts) == expected
+
+
+    # when the current user cannot access the given action
+    params = %{"id" => 2}
+    conn = conn(
+      %Plug.Conn{
+        private: %{phoenix_controller: Myproject.SampleController},
+        assigns: %{current_user: %User{id: 1}, canary_action: :someaction}
+      },
+      :get,
+      "/posts/2",
+      params
+    )
+    expected = Plug.Conn.assign(conn, :authorized, false)
+
+    assert authorize_controller(conn, opts) == expected
+
+    # when current_user is nil
+    params = %{"id" => 1}
+    conn = conn(
+      %Plug.Conn{
+        private: %{phoenix_controller: Myproject.SampleController},
+        assigns: %{current_user: nil, canary_action: :create}
+      },
+      :post,
+      "/posts",
+      params
+    )
+    expected = Plug.Conn.assign(conn, :authorized, false)
+
+    assert authorize_controller(conn, opts) == expected
+  end
```

Comment:
We should probably add a test for when authorization fails because a rule specified that the user could not access a given controller (as a regression test).